src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java

```diff
@@ -432,31 +432,42 @@ public void registerToolchains(boolean devDependency, Sequence<?> toolchainLabel
             named = true,
             positional = false,
             defaultValue = "False"),
+        @Param(
+            name = "isolated",
+            doc =
+                "If true, this usage of the module extension will be isolated from all other "
+                    + "usages, in particular those in other modules.",
+            named = true,
+            positional = false,
+            defaultValue = "False"),
       },
       useStarlarkThread = true)
   public ModuleExtensionProxy useExtension(
       String rawExtensionBzlFile,
       String extensionName,
       boolean devDependency,
+      boolean isolated,
       StarlarkThread thread) {
     hadNonModuleCall = true;
 
     String extensionBzlFile = normalizeLabelString(rawExtensionBzlFile);
 
     ModuleExtensionUsageBuilder newUsageBuilder =
         new ModuleExtensionUsageBuilder(
-            extensionBzlFile, extensionName, thread.getCallerLocation());
+            extensionBzlFile, extensionName, isolated, thread.getCallerLocation());
 
     if (ignoreDevDeps && devDependency) {
       // This is a no-op proxy.
       return newUsageBuilder.getProxy(devDependency);
     }
 
     // Find an existing usage builder corresponding to this extension.
-    for (ModuleExtensionUsageBuilder usageBuilder : extensionUsageBuilders) {
-      if (usageBuilder.extensionBzlFile.equals(extensionBzlFile)
-          && usageBuilder.extensionName.equals(extensionName)) {
-        return usageBuilder.getProxy(devDependency);
+    if (!isolated) {
+      for (ModuleExtensionUsageBuilder usageBuilder : extensionUsageBuilders) {
+        if (usageBuilder.extensionBzlFile.equals(extensionBzlFile)
+            && usageBuilder.extensionName.equals(extensionName)) {
+          return usageBuilder.getProxy(devDependency);
+        }
       }
     }
```

Comment:
How do we prevent a non-isolated instance from getting mapped to a previously created isolated instance?

Reply:
Thank you, that looks good to me.